core/tests/protractor_utils/editor.js

```diff
@@ -227,6 +228,152 @@ var expectCannotDeleteInteraction = function() {
+};
+
+// gadgetName must exist.
+var deleteGadget = function(gadgetName) {
+  element(by.css('.protractor-test-delete-' + gadgetName + '-gadget-icon'))
+    .click();
+  general.waitForSystem(); // wait for modal popup.
+  element(by.cssContainingText('.btn-danger', 'Delete Gadget')).click();
+};
+
+// Opens the editor modal for the specified gadget.
+// This method is a precursor to other methods like adjusting visibility.
+var openGadgetEditorModal = function(gadgetName) {
+  element(by.css('.protractor-test-edit-' + gadgetName + '-gadget')).click();
+};
+
+var saveAndCloseGadgetEditorModal = function() {
+  element(by.css('.protractor-test-save-gadget-button')).click();
+};
+
+// Check (activate) visibility for a given state.
+// openGadgetEditorModal must be called before this method.
+var checkGadgetVisibilityForState = function(stateName) {
+  element(by.cssContainingText(
+    '.protractor-test-state-visibility-checkbox-label',
+    stateName)).all(by.tagName('input')).then(function(items) {
+      items[0].isSelected().then(function(selected) {
+        if (!selected) {
+          items[0].click();
+        }
+      })
+    })
+};
+
+// Uncheck (disable) visibility for a given state.
+// openGadgetEditorModal must be called before this method.
+var uncheckGadgetVisibilityForState = function(stateName) {
+  element(by.cssContainingText(
+    '.protractor-test-state-visibility-checkbox-label',
+    stateName)).all(by.tagName('input')).then(function(items) {
+      items[0].isSelected().then(function(selected) {
+        if (selected) {
+          items[0].click();
+        }
+      })
+    })
+};
+
+// This can receive additional arguments for the gadget's customizationArgs.
+// Keep in mind this method applies to a gadget preview which may not be
+// fully operational.
+var expectGadgetPreviewToMatch = function(gadgetType, gadgetName) {
+  // Convert additional arguments to an array to send on.
+  var args = [element(
+    by.css('.protractor-test-' + gadgetName + '-gadget'))];
+  for (var i = 2; i < arguments.length; i++) {
+    args.push(arguments[i]);
+  }
+  gadgets.getGadget(gadgetType).
+    expectGadgetPreviewDetailsToMatch.apply(null, args);
+};
+
+var expectGadgetWithNameDoesNotExist = function(gadgetName) {
+  element.all(by.css('.protractor-test-' + gadgetName + '-gadget'))
+    .then(function(items) {
+      expect(items.length).toBe(0);
+    })
+};
+
+// PARAMETERS
+
+// This function adds a parameter change, creating the parameter if necessary.
+var addParameterChange = function(paramName, paramValue) {
+  element(by.css('.protractor-test-add-param-button')).click();
+
+  element(by.css('.protractor-test-param-changes-editor')).all(
+    by.tagName('input')).then(function(items) {
+      items[0].sendKeys(paramName, protractor.Key.ENTER);
+    });
+
+  element(by.css('.protractor-test-param-changes-editor')).all(
+    by.model('localValue.label')).then(function(items) {
```

Comment:
Not sure how best to implement this without using by.tagName('input') or by.model('localValue.label'). I followed this by.model('localValue.label') through param_change_editor.html to valueGeneratorEditor.html, to unicode_string_editor.html where I see both an input tag and localValue.label model on line 8. It appears at least one of these is using a 3rd party library called select2 that we didn't have an clean way to add protractor-test classes.

Related: it may not be apparent from the addParameterChange code, but getting protractor to work with the select2 pulldown was finicky and took hours. There are a number of stackoverflow posts on challenges using protractor with select2. I had to stop working on it one day and return to it later to get it working.  Open to any suggestions you have on a simple/effective solution here.